src/CountryHandler/Spain.php

```diff
@@ -29,25 +42,68 @@ final class Spain extends CountryHandler
     public const LENGTH = 9;
 
     /**
+     * @var array<string>
+     */
+    public const NIE = ['X', 'Y', 'Z'];
+
+    /**
+     * Spanish Natural Persons: DNI
+     * Foreigners with NIE.
+     *
      * @var string
      */
-    public const PATTERN_1 = '\\d{8}[a-zA-Z]';
+    public const PATTERN_1 = '(^[XYZ\d]\d{7})([' . self::CONTROL_1 . ']$)';
 
     /**
+     * Non-resident Spaniards without DNI
+     * Resident Spaniards under 14 without DNI
+     * Foreigners without NIE
+     * Legal entities (companies, organizations, public entities, ...).
+     *
      * @var string
      */
-    public const PATTERN_2 = '[XYZKLMxyzklm]\\d{7}[a-zA-Z]';
+    public const PATTERN_2 = '(^[ABCDEFGHJKLMNPQRSUVW])(\d{7})([' . self::CONTROL_2 . '\d]$)';
 
     /**
-     * @var array<int, string>
+     * Return checksum char for Spanish TIN.
+     *
+     * @param string $tin
+     * The TIN without Country indicative ('ES')
+     * @param null|bool $digit
+     * Optional: for Non-Natural Persons TIN forces return checksum char as digit 0-9
+     *
+     * @return null|string
+     * Return checksum char or null on failure
      */
-    private static $tabConvertToChar = [
-        'T', 'R', 'W', 'A', 'G',
-        'M', 'Y', 'F', 'P', 'D',
-        'X', 'B', 'N', 'J', 'Z',
-        'S', 'Q', 'V', 'H', 'L',
-        'C', 'K', 'E',
-    ];
+    public static function getChecksum(string $tin, ?bool $digit = null): ? string
```

Comment:
1. Since this seems to be the most critical part, I will ask you if there is already an existing algorithm that implement the same `getChecksum` function somewhere else? Do you have the documentation for that specific part?
2. Why is this a `public static` function?

Reply:
1. The Spanish TIN has two algoritms:
  a. [Natural Persons with DNI or NIE](https://github.com/fawno/tin/blob/master/docs/TIN-Country_Sheet_ES.md#natural-persons-with-dni-or-nie-1)
  b. [Natural Persons without DNI or NIE and Non-Natural Persons](https://github.com/fawno/tin/blob/master/docs/TIN-Country_Sheet_ES.md#natural-persons-without-dni-or-nie-and-non-natural-persons)
  I wrote a documentation with the official references I found.
2. I think it is interesting to be able to calculate the control digit. There are still many people in Spain who do not know the control digit but do know their DNI number (TIN without checksum character).
3. It is a static method because it allows access to the calculation of the check digit without having to create an object that we may not need.

Reply:
1. I'll review this
2. Fair enough...but... this library is not about calculating the control digit, it's about verifying if a TIN is valid or not.
3. There is no need to expose this to the public nor making it static, I would kindly ask you to keep that private in the file.

Reply:
To verify the check digit you have to calculate it. I think it is interesting to expose the method of calculating the check digit.
Having the method accessible allows other projects to create error correction routines. And it is just one of the things that comes to mind.

I like the idea of ​​having the method accessible as static. But if you maintain your opinion I will change it to a private method.

Reply:
I understand your point of view, and I do understand that having such a function would be useful, but this is not the purpose of the library and I don't want to endorse that responsibility through the library, I prefer to delegate that to the end user for the moment. The `CountryHandler` has a well-defined interface and I prefer to avoid modifying it. Also I don't want to add random public static function in there too. So to fully answer your question, yes I prefer to avoid static function, and I'd like to keep it private.
We can re-evaluate that option in the next major version of the library for sure.